internal/pkg/agent/cmd/upgrade_test.go

```diff
@@ -67,6 +82,426 @@ func TestUpgradeCmd(t *testing.T) {
 		close(upgradeCh)
 		// this makes sure all client assertions are done
 		<-clientCh
+		c.Disconnect()
+	})
+
+	t.Run("fail if fleet managed and unprivileged", func(t *testing.T) {
+		var wg sync.WaitGroup
+		// Set up mock TCP server for gRPC connection
+		tcpServer, err := net.Listen("tcp", "127.0.0.1:")
```

Comment:
Only investigate this if you want to, but I think these tests can be simplified because you don't actually need a server to test gRPC server implementations. You can just invoke the `mock.Upgrade()` RPC implementation directly without needing a network transport to connect you to it.

You would probably have to make a test client type that wraps the server type, so that calling `client.Upgrade()` just immediately calls `mock.Upgrade()`.

Reply:
This makes much more sense. I just looked at the already implemented test case and followed the same approach for the new ones that I added. Comparing the test cases I added to the one that was already implemented, I see that the new ones don't necessarily need an actual server running as we are not testing for connection interruptions. Will look into this one; however, it is not a priority for me. I may open a new PR to refactor these test cases.

Reply:
Updating this now actually. Instead of mock server, I am using mock client